### ISSUE TITLE
Cannot run cells until a kernel had been auto selected (with push model)

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode#122344]
With the push model running cells is now slow:
I know need to wait for the kernel to get auto selected.
However previously this was the workflow:
* As soon as notebook opened I could hit the run button

As a user I don't know nor do i care about auto selection of kernels..all I know is I want to run a cell, hence i should be hit the run button to queue that cell. If a kernel is not available then it should prompt the user for a kernel and then run it, or when a preferred kernel is available it should pick that and run it.

Basically, notebooks are now slow in getting to a ready state,  to accept user requests to run a cell.

@jrieken @rebornix  /cc

**Comment by jrieken** [comment]
> Basically, notebooks are now slow in getting to a ready state, to accept user requests to run a cell.

Since we are not asking for kernels/controller anymore this is now 100% under your control. The faster your extension creates a controller, the better this gets. 

This is not slower, it might feel slower but the only thing that has changed is that there is no run-button until a controller is available. That's the inherit characteristics of the push model and nothing we can change

**Comment by jrieken** [comment]
>  If a kernel is not available then it should prompt the user for a kernel and then run it, 

While we could show a run button which isn't doing anything but waiting we cannot "invent" kernels to pick from. What should a user answer is such a kernel prompt?

> or when a preferred kernel is available it should pick that and run it.

This will happen kinda happen - as soon as any kernel is available and run happens. What isn't happening is that we don't show the run button immediately. We could but it has potential for confusion because fast but unwanted controllers can win against slow but wanted controllers.

**Comment by DonJayamanne** [comment]
I understand the technical limitations, however as a user see l this s sense to provide a broken ux.
The previous approach seemed better and more intuitive.

**Comment by jrieken** [comment]
As an extension author you should see this as challenge to make your controller creation as fast as possible

**Comment by DonJayamanne** [comment]
> As an extension author you should see this as challenge to make your controller creation as fast as possible

Agreed, however there are limitations at our end (performing io operations, file/for reads and process need to be launched). And I'd hope vscode could queue the user request (to run) as it did.
The extension could be slow for a number of reasons.

**Comment by jrieken** [comment]
> And I'd hope vscode could queue the user request (to run) as it did.

Sure, but the question is for how long should we queue them? A few seconds? Until the associated controller comes along? Until any controller comes along?

**Comment by jrieken** [comment]
I am happy to experiment with this but I want to avoid giving an automatic preference to a fast controller...

**Comment by DonJayamanne** [comment]
> h this but I want to avoid giving an automatic preference to a fast controller...

Agreed. Challenges of push model (I guess pull has the same issues).
The UX is similar to code completion.
When I open a Python file and ask for completions, it seems VSCode queues these messages until a language server is ready, hence I'd expect the same (as a user).

**Comment by DonJayamanne** [comment]
One solution I have is for the extension to register a bogus preferred controller, this way VS Code has a preferred controller. & we (extension) can queue the cell execution until we have the real controllers ready.

Thats definitely doable today, but would require a lot of work arounds at our end.
I'll check with the rest of team, though this feels like a hacky work around.

**Comment by DonJayamanne** [comment]
> jupyter kernels it has been using during the last session and optimistically create controllers for them again

Sorry, I was thinking more about first time use.
& Yes for the second visits we can do what you suggested.
But either way, we'd need a temporary (what i originally referred to as bogus) controller

Lets leave this issue open, i'd like to look into this further at our end. Thanks

**Comment by DonJayamanne** [comment]
@jrieken i think we need to fix this in VS Code.
Our PMs have run into this issue as well, I've had a chat with @rebornix as well.
I think VS Code should do something here, else its a poor UX.

Can't really find a good solution here:
* VS Code will look for the first available preferred kernel and run
* VS Code could wait for n seconds & then prompt the user to select a kernel.
	* This feels wrong at so many levels, in Jupyter, we could be slow at finding the kernels (we might have to wait for Python extension to activate, etc). in such a case it could take a few seconds.

Personally I feel the pull model solved this better, and the push has complicated this scenario.

Basically the problem is:
* Open a notebook
* Ther'es no run button
* Wait for 3 seconds and magically the button appears.
To me this is like auto completion, if the user hits `ctrl+space` vscode should queue that request and when an LS is ready return the results & display them to the user.

We've run into this issue (run not being available) a few times & its a little annoying.

**Comment by jrieken** [comment]
> Personally I feel the pull model solved this better, and the push has complicated this scenario.

I am sorry but you missed your shot 5 weeks ago, https://github.com/microsoft/vscode/issues/119949

<img width="931" alt="Screenshot 2021-04-30 at 09 19 45" src="https://user-images.githubusercontent.com/1794099/116661818-4bb05280-a995-11eb-9fff-0c98788e0fcd.png">

So, please provide constructive feedback

**Comment by DonJayamanne** [comment]
> So, please provide constructive feedback

Sorry you felt this is not constructive. I'm not saying that the push model is not good compared to pull. Just saying that the pull solved this particular problem better.
I think you're assuming something entirely different here.

>  am sorry but you missed your shot 5 weeks ago

Right now I'm not sure what you want me to do. I'm just providing feedback on a problem with some proposed solutions (I'd consider that constructive).

**Comment by jrieken** [comment]
> I'm just providing feedback on a problem with some proposed solutions

You feedback is not actionable. You say "Wait for 3 seconds and magically the button appears." but what does that mean?

* Does it take 3 seconds for every notebook or the very first notebook?
* Why does it take 3 seconds? Running `jupyter kernelspec list` takes 150ms on my machine.
* Is that an average number, a median, p95, a random value?

I have no clue what you are suggesting? Do want us to go back to the pull model? Should explore UI ideas that indicate that kernels are about to come? Do you want API to drive such a UI or should we be optimistic and try this without API?

**Comment by DonJayamanne** [comment]
> Why does it take 3 seconds? Running jupyter kernelspec list takes 150ms on my machine.

You're assuming that's all we do. That's not just it.we look for interpreters as well, we try to match interpreters with kernels, get version info etc. And we need to wait for python extension to activate (that's what gives us information about python). 3s is an arbitrary number I pulled from no where. To indicate it could be anything.

> random value

Yes, that's right.

> Should explore UI ideas that indicate that kernels are about to com

I'm trying to start a conversation to look into this problem with possible solutions. I'm  suggesting something and highlighting problems with those suggestions.

Last time I suggested we register a bogus controller. I'd like is to go back and look/think about a more general solution, a more genetic solution, beaded on my discussions with Peng he to felt this (current) was not the best ux. So I'm trying to start a conversation here.

I.e. I want to get the ball rolling with the discussion.

**Comment by jrieken** [comment]
Yes, I am happy to brain storm about UI ideas but this is a tricky problem because it is just guess work on our end. 

Therefore, I wanna make sure that you have done everything to create controllers as fast as possible. What do you need to do to invoke the `createNotebookController`-function and what can you do afterwards? Note that everything except for `id`, `viewType`, and `preloads` can be changed afterwards. Is `preloads` the problem? Let's think how you can provide that later. Do you really need to match interpreters with kernels _before_ creating the corresponding controller? I wanna go into depth first, then consider UI tricks. Let looks at the arguments of `createNotebookController`

* `id` - What do you need to do to come up with a stable id? Is that derived from the location of disk?
* `viewType` - This is constant, right?
* `label` - When do you know the label? Can you use a temporary label and update it quickly after?
* `handler` - This optional, you can also use a handler that represents the aftermath that is needed before execution is possible
* `preloads` - This might be a challenge. Should we have a `addPreload`-function instead?

So, please shed some light. Which of those three readonly properties are expensive to compute and where is the trouble?

**Comment by jrieken** [comment]
to reveal my strategy here: show me that you are creating controllers the fastest possible way and then we build some "progress" UI that nobody ever sees.

**Comment by DonJayamanne** [comment]
Here's what we do as part of getting a list of controllers (kernels):
* Find all kernel specs on disc (this is a simple search on disc)
* Find all python interpreters on disc (get this from Python extension, this is one of the slowed parts)
* For each python interpreter look for specific kernels
	* 	E.g. if Python is located in `C:/Users/XYZ/ABC/PythonEnv/`, then look for Jupyter kernel specs in `C:/Users/XYZ/ABC/PythonEnv/shared/jupyter/kernels`
* For each kernel spec try to find the corresponding Python interpreter
	* Possible there's a 1-1 match for them.
	* If ther'es a match, then we just hide (exclude) these kernels specs
* What ever remains
	* Display those kernel specs
	* Display those Python interpreters


Hence the slowest part is Python Interpreter (Environment) discovery

Yes, when loading VS Code for the second time we could try to load these from cache.
I'll also check our telemetry to see how slow this entire process is at our end.

**Comment by jrieken** [comment]
> Find all kernel specs on disc (this is a simple search on disc)

Do you create controllers on the go? Like, do you call `createNotebookController` while finding kernel specs or do you wait for all?

**Comment by DonJayamanne** [comment]
We wait for all, 
We have to wait for all kernels + interpreters have been found (those operations are async & slowest is interpreter discovery via Python extension).
The matching & removal of duplicates is synchronous & not slow.

**Comment by jrieken** [comment]
> We wait for all,

Isn't this contrary to what you said here: https://github.com/microsoft/vscode/issues/116504#issuecomment-783622239:

> 90% of the time, user will have to wait for 10 seconds to run a notebook just because VS Code is fetching all of the kernels, when all it needs is just the preferred kernel.

The push model allows you bring in kernels as you discover them, giving you all the control, but it seems you are mimicking the old behavior (pull model) where the slowest part of the kernel discovery determines the overall experience

**Comment by DonJayamanne** [comment]
> Isn't this contrary to what you said here

Sorry I was wrong there 

> but it seems you are mimicking the old behavior (pull model) where the slowest part of the kernel discovery determines the overall experience

As mentioned we need to get all kernels and so interpreters and me the two.
If it's an existing notebook (user opened previously on the same computer), then we can try get the preferred faster as we could cache things and we know exactly what was users and see don't need to go through all the kennels and interpreters.

There's no other way to do this, we must wait for all  to get the preferred when it's a new notebook (untitled or one we have more seen before), new work space or a new computer... 

Ie getting preferred in our case can only be done when we have the entire list, is at pick the best from the lot. This is by design, not due to a kernel pull api design.

**Comment by jrieken** [comment]
> Sorry I was wrong there , please see below 

🤣 Oh, so the whole new API is based on a mistake? Over and out...

> There's no other way to do this, we must wait for all to get the preferred when it's a new notebook

Can you also explain why that is?

**Comment by DonJayamanne** [comment]
> the whole new API is based on a mistake? Over and out

If a user opens an existing notebook, we can optimize by just providing the preferred kernel. That's one scenarios I was referring to. It's not a blanket statement for ALL scenarios. 

Looks like you have made some  assumption and I feel like  you are trying to use me or something I said as a scapegoat for some poor api or families in the api. 

If it was all a mistake then can at go back to the api we had, sure it's a method change, if we changed the whole api once  we can do it again.. 

> Can you also explain why that is?

I have tried to explain this. I think it's best at discuss this on a teams meeting. I'll include the rest of my team Jeff as well to existing how at fetch the kennels and interpreters. Please feel free to schedule a meeting to discuss this with @rchiodo @greazer

**Comment by jrieken** [comment]
I am sorry but I am still confused and I sense a lot of emotion so maybe it is better if @rchiodo or @IanMatthewHuff can chime in. 

* Is it that Jupyter disagrees with the push model of kernels and does Juptyer want the return of the pull model where VS Code asks all extensions for all kernels and where the slowest operation determines the overall performance, or
* do you want so some kind of API that allows you say "I am Jupyter and I am looking for kernels"? Something like `window.withProgress` but geared towards notebooks? 


> There's no other way to do this, we must wait for all to get the preferred when it's a new notebook

I am asking this question again - out of curiosity and not because I want to annoy you. Do new (as in untitled) notebooks have preferred kernels? Do you wait for all kernels as a "workaround" to the current UI/flow where a controller appear as soon as it is created?

**Comment by rchiodo** [comment]
Not sure if push works for us or not. Probably need to think about it more.

The problem we have is that we have to fetch all of the interpreters before we can return any kernels. Kernels are not just the kernelspec.json on disk, but rather that and the closest match for an interpreter that we need to use to run that kernel. 

Maybe we could rework things to return a list right away and then update them later? Meaning just return the name of a kernel say python 3 (assuming the notebook specifies that) and then generate the true list later. Not sure if the API allows subsequent updates or not.

> Do new (as in untitled) notebooks have preferred kernels?

Untitled notebooks would have a preferred kernel (the Python 3 I mentioned above). But the same thing applies to them. We have to figure out the interpreter that goes with it. They're faster as they can just use the active interpreter but on first launch just computing that can take 10 seconds (especially since it loads the python extension).

> Do you wait for all kernels as a "workaround" to the current UI/flow where a controller appear as soon as it is created?

I believe we're waiting for all kernels because we can't return something runnable until all of the interpreters are found.

Does the API allow us to update the current kernel after we compute the relevant data? Cause I think if it did, we could just generate an immediate dummy kernel that matches the kernel specified in a notebook (or 'python 3' if untitled) and then update it's relevant data later.

**Comment by jrieken** [comment]
Thanks for the quick response!

> Meaning just return the name of a kernel say python 3 (assuming the notebook specifies that) and then generate the true list later. Not sure if the API allows subsequent updates or not.

Yes, you can update everything except for `id`, `viewType`, and `preloads` (see https://github.com/microsoft/vscode/issues/122344#issuecomment-829981323). For the `preloads` I am basically waiting for feedback if that should be relaxed or not. It is easy to add preloads later but removing is more tricky (because it requires a reload of the renderer) and therefore it is currently readonly.

> Cause I think if it did, we could just generate an immediate dummy kernel that matches the kernel specified in a notebook (or 'python 3' if untitled) and then update its relevant data later.

That's exactly what I am hoping for: create kernels as soon as possible (so that the UI shows the kernel and not progress bars) and then enrich it with more data.

**Comment by rchiodo** [comment]
I believe right now the id contains the interpreter data in it. It's how we refer to them later. Maybe we can use a guid instead and have an internal mapping to another id. 

@DonJayamanne, @IanMatthewHuff does this sound reasonable?

- Read notebook metadata
- Check not somebody else's kernel (like .NET)
- If our kernel, create kernel with metadata name, id set to guid and default preloads (preloads are the same for all of our kernels). Set some internal property on the kernel that specifies pending
- Continue to read the interpreters and kernelspecs in the background. 
- When done
   - Update map of guids to internal id
   - Update pending kernel with the necessary data for execution
   - Send full list of kernels to VS code.

**Comment by DonJayamanne** [comment]
Sure, basically we're just adding a dummy one, waiting for everything to complete & then updating that dummy item with real inforamtion.
As for the ID, i don't think we use that ID ingernally, hence yes, we can always use a GUID.

**Comment by DonJayamanne** [comment]
Yes, that will work.

**Comment by DonJayamanne** [comment]
Closing this issue as we're going to look into the solution of having a dummy kernel.
